File: java/PopulatingNextRightPointersInEachNodeII.java
import java.util.ArrayList;

public class PopulatingNextRightPointersInEachNodeII {
    public void connect(TreeLinkNode root) {
        TreeLinkNode prev = null;
        TreeLinkNode post = null;
        TreeLinkNode curr = root;
        for ( ; curr != null; ) {
            prev = null;
            post = null;
            for ( ; curr != null; curr = curr.next) {
                if (curr.left == null && curr.right == null) {
                    continue;
                }
                if (post == null) {
                    if (curr.left != null) {
                        post = curr.left;
                    } else {
                        post = curr.right;
                    }
                }
                if (curr.left != null) {
                    if (prev != null) {
                        prev.next = curr.left;
                    }
                    prev = curr.left;
                }
                if (curr.right != null) {
                    if (prev != null) {
                        prev.next = curr.right;
                    }
                    prev = curr.right;
                }
            }
        }
    }
}

File: python/populating_next_right_pointers_in_each_node_ii.py
#!/usr/bin/env python3

class TreNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
        self.next = None

class Solution:
    # @param root, a tree node
    # @return nothing
    def connect(self, root):
        ptr = root
        while ptr:
            first = None
            prev = None
            while ptr:
                if ptr.left:
                    if first is None:
                        first = ptr.left
                    if prev:
                        prev.next = ptr.left
                    prev = ptr.left
                if ptr.right:
                    if first is None:
                        first = ptr.right
                    if prev:
                        prev.next = ptr.right
                    prev = ptr.right
                ptr = ptr.next
            ptr = first
